### PR TITLE
Use high resolution credits on new contract accounts.

### PR DESCRIPTION
**Pull request by DanielVF**
We are targeting [increasing the credits resolution for OUSD](https://github.com/OriginProtocol/origin-dollar/issues/590). Our immediate motivation is Uniswap V3 support, which requires balance-exact transfers in order to work. Upgrading all accounts will require substantial work.

However, it turns out that a new, non-rebasing account has its own independent creditsPerToken ratio, and has no effect on OUSD's bookkeeping values This means that it is trivial to set the new account to start at a high resolution creditsPerToken ratio, and as a result transfers in and out of that account will be balance exact.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [x] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/contracts/token/OUSD.sol [hunk 1]
```diff
@@ -382,12 +382,22 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
...
     function _ensureRebasingMigration(address _account) internal {
         if (nonRebasingCreditsPerToken[_account] == 0) {
-            // Set fixed credits per token for this account
-            nonRebasingCreditsPerToken[_account] = rebasingCreditsPerToken;
-            // Update non rebasing supply
-            nonRebasingSupply = nonRebasingSupply.add(balanceOf(_account));
-            // Update credit tallies
-            rebasingCredits = rebasingCredits.sub(_creditBalances[_account]);
+            if (_creditBalances[_account] == 0) {
+                // Since there is no exisiting balance, we can directly set to
```
Comment (tomlinton) [inline contracts/contracts/token/OUSD.sol:386]: Typo in comment

**Comment by tomlinton** [comment]
Great idea, I was a bit worried about opt ins/outs but running through those it looks like it'll be fine.

**Review by franckc — COMMENTED** [review]
Logic looks good. Neat trick :)

Couple questions:
 1. remind me, how did we pick the 1e27 resolution? (just to double check this is what we want).
 2. is there any chance this make fixing the rest of the accounts (which I assume is the majority) more challenging since we'll have to check if an account uses high resolution or not?

Let's add a migration file. Either in this PR or in a separate PR. Up to you.

File: contracts/contracts/token/OUSD.sol (hunk 1 above)
Comment (DanielVF) [reply contracts/contracts/token/OUSD.sol]: Fixed. Thanks!

**Comment by DanielVF** [comment]
1e27 is a standard "RAY" in DAI world, and is used by AAVE v2 tokens as their standard internal measurement. Since our public balances is 1e18, this gives us roughly 1e9 - 1e1 = 1e8, or letting OUSD grow 100,000,000x before we run into the same resolution problems as we had before. At 20% growth per year, this would last for 101 years. 

The other side is how much total supply this would support. 

- 2^256 is roughly 1e77.
- 1e77 (underlying) - 1e27 (decimal point) is 1e50 dollars
- World circulating currency and bank accounts is roughly 37 trillion or 1e14
- If the dollar inflation averages 20% per year, for 101 years, that's 1e8
- So OUSD would be able to support 1e50 - 1e14 (world currency) - 1e8 (inflation) = 1e28 or 10 Octillions times the world's money for 100 years.

So a little bit overkill on that side, and we might honestly be able to split the difference and do our internal resolution at 1e40, I just went with 1e27 since it's common for this kind of thing.

**Comment by franckc** [comment]
Thanks for the details about 1e27 resolution :) Love the back of the envelope calculation. Clearly shows we should be good for a while. Next will look at the new migration...

File: contracts/utils/deploy.js [hunk 2]
```diff
@@ -236,6 +236,56 @@ const sendProposal = async (proposalArgs, description) => {
   log("Done");
 };
 
+async function deployment(deployName, fn) {
```
Comment (franckc) [inline contracts/utils/deploy.js:239]: Maybe also add an optional dependencies (list of strings) argument.
And have something like
```
if (dependencies && dependencies.length > 0) {
  main.dependencies = dependencies.
}
```

Comment (franckc) [reply contracts/utils/deploy.js:239]: nit: add jsdoc for the new 'deployment' method

File: contracts/utils/deploy.js [hunk 3]
```diff
@@ -245,4 +295,5 @@ module.exports = {
   executeProposal,
   executeProposalOnFork,
   sendProposal,
+  // deployment
```
Comment (franckc) [inline contracts/utils/deploy.js:298]: uncomment?

**Review by franckc — APPROVED** [review]
LGTM

I like the refactor of the deployment code to reduce the amount of boilerplate code we havae to write moving forward. Nice! :)

File: contracts/utils/deploy.js (hunk 2 above)
Comment (DanielVF) [reply contracts/utils/deploy.js]: Thanks, added.

File: contracts/utils/deploy.js (hunk 3 above)
Comment (DanielVF) [reply contracts/utils/deploy.js]: Fixed, thanks!

**Pull request merged**